plugins/ocp/ocp-nvme.c

```diff
@@ -637,6 +828,454 @@ static int eol_plp_failure_mode(int argc, char **argv, struct command *cmd,
+		size    = le16_to_cpu(logheader->DataArea1LastBlock);
+		break;
+
+	case 2:
+		offset  = TELEMETRY_HEADER_SIZE
+				+ (le16_to_cpu(logheader->DataArea1LastBlock) * TELEMETRY_BYTE_PER_BLOCK);
+		size    = le16_to_cpu(logheader->DataArea2LastBlock)
+				- le16_to_cpu(logheader->DataArea1LastBlock);
+		break;
+
+	case 3:
+		offset  = TELEMETRY_HEADER_SIZE
+				+ (le16_to_cpu(logheader->DataArea2LastBlock) * TELEMETRY_BYTE_PER_BLOCK);
+		size    = le16_to_cpu(logheader->DataArea3LastBlock)
+				- le16_to_cpu(logheader->DataArea2LastBlock);
+		break;
+
+	default:
+		break;
+	}
+
+	if (size == 0) {
+		printf("Telemetry %s Area %d is empty.\n", featurename, data_area);
+		return err;
+	}
+
+	// Get Data Block
+	char dumpname[FILE_NAME_SIZE] = {0,};
+
+	snprintf(dumpname, FILE_NAME_SIZE,
+					"Telemetry_%s_Area_%d", featurename, data_area);
+	err = extract_dump_get_log(dev, dumpname, filename, sn, size * TELEMETRY_BYTE_PER_BLOCK,
+			TELEMETRY_TRANSFER_SIZE, nsid, tele_type,
+			0, offset, rae);
+
+	return err;
+}
+
+static int ocp_telemetry_log(int argc, char **argv, struct command *cmd,
+			      struct plugin *plugin)
+{
+	struct nvme_dev *dev;
+	int err = 0;
+	const char *desc = "Retrieve and save telemetry log.";
+	const char *type = "Telemetry Type; 'host[Create bit]' or 'controller'";
+	const char *area = "Telemetry Data Area; 1 or 3";
+	const char *file = "Output file name with path;\n"
+			"e.g. '-o ./path/name'\n'-o ./path1/path2/';\n"
+			"If requested path does not exist, the directory will be newly created.";
+
+	__u32  nsid = NVME_NSID_ALL;
+	struct stat nvme_stat;
+	char sn[21] = {0,};
+	struct nvme_id_ctrl ctrl;
+	bool is_support_telemetry_controller;
+
+	int tele_type = 0;
+	int tele_area = 0;
+
+	struct config {
+		char *type;
+		int area;
+		char *file;
+	};
+
+	struct config cfg = {
+		.type = NULL,
+		.area = 0,
+		.file = NULL,
+	};
+
+	OPT_ARGS(opts) = {
+		OPT_STR("telemetry_type", 't', &cfg.type, type),
+		OPT_INT("telemetry_data_area", 'a', &cfg.area, area),
+		OPT_FILE("output-file", 'o', &cfg.file, file),
+		OPT_END()
+	};
+
+	err = parse_and_open(&dev, argc, argv, desc, opts);
+	if (err)
+		return err;
+
+	err = fstat(dev_fd(dev), &nvme_stat);
+	if (err < 0)
+		return err;
+
+	if (S_ISBLK(nvme_stat.st_mode)) {
+		err = nvme_get_nsid(dev_fd(dev), &nsid);
+		if (err < 0)
+			return err;
+	}
+
+	err = nvme_identify_ctrl(dev_fd(dev), &ctrl);
+	if (err != 0) {
+		return err;
+	}
+
+	get_serial_number(&ctrl, sn);
+
+	is_support_telemetry_controller = ((ctrl.lpa & 0x8) >> 3);
+
+	if (!cfg.type && !cfg.area) {
+		tele_type = TELEMETRY_TYPE_NONE;
+		tele_area = 0;
+	} else if ((!cfg.type + !cfg.area) == 0) {
+		if (!strcmp(cfg.type, "host0"))
+			tele_type = TELEMETRY_TYPE_HOST_0;
+		else if (!strcmp(cfg.type, "host1"))
+			tele_type = TELEMETRY_TYPE_HOST_1;
+		else if	(!strcmp(cfg.type, "controller"))
+			tele_type = TELEMETRY_TYPE_CONTROLLER;
+
+		tele_area = cfg.area;
+
+		if (!((((tele_type == TELEMETRY_TYPE_HOST_0)
```

Comment:
Indentation error and seems used uncessary `(` and `)` as complicated conditons and hard to understand.
```diff
- 		if (!((((tele_type == TELEMETRY_TYPE_HOST_0)
- 				|| (tele_type == TELEMETRY_TYPE_HOST_1))
- 				&& ((tele_area == 1) ||  (tele_area == 3)))
- 				|| (tele_type == TELEMETRY_TYPE_CONTROLLER && tele_area == 3))) {
+  		if (!(((tele_type == TELEMETRY_TYPE_HOST_0 || tele_type == TELEMETRY_TYPE_HOST_1) &&
+ 		    (tele_area == 1 || tele_area == 3)) || tele_type == TELEMETRY_TYPE_CONTROLLER &&
+ 		    tele_area == 3)) {
```

Reply:
How aoubt the suggesution below?
```diff
- 		if (!(((tele_type == TELEMETRY_TYPE_HOST_0 || tele_type == TELEMETRY_TYPE_HOST_1) &&
- 		    (tele_area == 1 || tele_area == 3)) || tele_type == TELEMETRY_TYPE_CONTROLLER &&
- 		    tele_area == 3)) {
+  		if (tele_area != 1 || tele_area != 3 ||
+  		    tele_type == TELEMETRY_TYPE_CONTROLLER && tele_area != 3)
```

Reply:
I changed it to
```
if ((tele_area != 1 && tele_area != 3) ||
			(tele_type == TELEMETRY_TYPE_CONTROLLER && tele_area != 3))
```
instead.

Reply:
I see. Thank you.